src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java

```diff
@@ -0,0 +1,429 @@
+   * <ul>
+   * <li>flush is called, or</li>
+   * <li>the linger timeout passes, or</li>
+   * <li>there are sufficient records to fill a batch</li>
+   * </ul>
+   */
+  private synchronized boolean canSubmit(long elapsedMs) {
+    return !unsentRecords.isEmpty()
+           && (flushRequested || elapsedMs >= lingerMs || unsentRecords.size() >= batchSize);
+  }
+
+  /**
+   * Start concurrently creating and sending batched requests using the client.
+   */
+  public void start() {
+    farmer.start();
+  }
+
+  /**
+   * Initiate shutdown.
+   *
+   * Pending buffered records are not automatically flushed, so call {@link #flush(long)} before this method if this is desirable.
+   */
+  public void stop() {
+    log.trace("stop");
+    stopRequested = true;
+    synchronized (this) {
+      // shutdown the pool under synchronization to avoid rejected submissions
+      executor.shutdown();
+      notifyAll();
+    }
+  }
+
+  /**
+   * Block upto {@code timeoutMs} till shutdown is complete.
+   *
+   * This should only be called after a previous {@link #stop()} invocation.
+   */
+  public void awaitStop(long timeoutMs) {
+    log.trace("awaitStop {}", timeoutMs);
+    assert stopRequested;
+    try {
+      if (!executor.awaitTermination(timeoutMs, TimeUnit.MILLISECONDS)) {
+        throw new ConnectException("Timed-out waiting for executor termination");
+      }
+    } catch (InterruptedException e) {
+      throw new ConnectException(e);
+    } finally {
+      executor.shutdownNow();
+    }
+  }
+
+  /**
+   * @return whether {@link #stop()} has been requested
+   */
+  public boolean isStopping() {
+    return stopRequested;
+  }
+
+  /**
+   * @return whether any task failed with an error
+   */
+  public boolean isFailed() {
+    return error.get() != null;
+  }
+
+  /**
+   * @return {@link #isTerminal()} or {@link #isFailed()}
+   */
+  public boolean isTerminal() {
+    return isStopping() || isFailed();
+  }
+
+  /**
+   * Throw a {@link ConnectException} if {@link #isStopping()}.
+   */
+  public void throwIfStopping() {
+    if (stopRequested) {
+      throw new ConnectException("Stopping");
+    }
+  }
+
+  /**
+   * Throw the relevant {@link ConnectException} if {@link #isFailed()}.
+   */
+  public void throwIfFailed() {
+    if (isFailed()) {
+      throw error.get();
+    }
+  }
+
+  /**
+   * {@link #throwIfFailed()} and {@link #throwIfStopping()}
+   */
+  public void throwIfTerminal() {
+    throwIfFailed();
+    throwIfStopping();
+  }
+
+  /**
+   * Add a record, may block upto {@code timeoutMs} if at capacity with respect to {@code maxBufferedRecords}.
+   *
+   * If any task has failed prior to or while blocked in the add, {@link ConnectException} will be thrown with that error.
+   */
+  public synchronized void add(R record, long timeoutMs) {
+    final long addStartTimeMs = time.milliseconds();
+    for (long elapsedMs = time.milliseconds() - addStartTimeMs;
```

Comment:
This code is called in a pretty tight loop since it is invoked for every record. Might be worth trying to avoid at least 1 of the 2 mandatory calls to `time.milliseconds()` as those calls aren't always cheap (we've seen performance problems w/ it in `KafkaProducer` before). You could probably even get rid of both if you check the conditions for adding a record before even recording the start time.